yt/chyt/server/query_analyzer.cpp

```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
     OptimizedQueryProcessingStage_ = DB::QueryProcessingStage::Complete;
 }
 
+EReadInOrderMode TQueryAnalyzer::InferReadInOrderMode() const
+{
+    auto selectQuery = QueryInfo_.query->as<DB::ASTSelectQuery>();
+
+    // Read in order is forbidden in JOINs and queries with aggregation.
+    // It might be useful in some of these cases, but that is a question for another day.
+    if (Join_ || selectQuery->groupBy() || selectQuery->having() || selectQuery->window() || selectQuery->limitBy()) {
+        return EReadInOrderMode::None;
+    }
+
+    // Read in order makes no sense without requested order.
+    // Native CH optimizations will do the trick.
+    if (!selectQuery->orderBy()) {
+        return EReadInOrderMode::None;
+    }
+
+    // Read in order makes no sense without any limits specified.
+    // The whole table will probably be read anyway.
+    if (!selectQuery->limitLength() && !getContext()->getSettingsRef().limit) {
+        return EReadInOrderMode::None;
+    }
+
+    YT_VERIFY(YtTableCount_ == 1);
+    const auto& schema = Storages_[0]->GetSchema();
+
+    // If the underlying table is not sorted we cannot help in any way.
+    if (!schema->IsSorted()) {
+        return EReadInOrderMode::None;
+    }
+
+    auto commonDirection = EReadInOrderMode::None;
+
+    // Columns from the ORDER BY clause must form a prefix of the table's primary key.
+    if (std::ssize(selectQuery->orderBy()->children) > schema->GetKeyColumnCount()) {
+        return EReadInOrderMode::None;
+    }
+
+    std::vector<TString> floatKeyColumnsWithoutExplicitNullsDirection;
+
+    for (const auto& [columnIndex, orderByElementAst] : Enumerate(selectQuery->orderBy()->children)) {
+        auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+        // Judging by CH code and experiments, ASTOrderByElement always has a single child.
+        if (std::ssize(orderByElement->children) != 1) {
+            THROW_ERROR_EXCEPTION("Query ORDER BY ast is malformed; "
+                "this is a bug; please contact cluster administrators");
+        }
+
+        // The name can contain functions. While it is possible to support tuples
+        // and even monotonic functions, I do not think there is much use.
+        auto columnName = orderByElement->children.front()->getColumnName();
```

Comment:
Need to check that the orderByElement is a `column` somehow before accessing it.

We don't want to break queries with an expression in the `order by` clause.

Reply:
We shouldn't? A few lines below we compare this name to the name of the corresponding column (by index) in the schema. This should never be true for expressions, so we won't enable the optimization. There is one test for that.

Reply:
I expected it to throw an error if the `orderByElement` is not a column literal. But yeah, since the tests work as expected, it's not the case.

Anyway, imho, an explicit check that the element is a literal would show your intent clearer in the code.

Reply:
Changed it to a cast to `ASTIdentifier`. It is even better, because it works properly with aliases introduced by `WITH`!